src/cryptonote_core/service_node_quorum_cop.cpp

```diff
@@ -178,6 +181,17 @@ service_node_test_results quorum_cop::check_service_node(
             log::info(logcat, "Service Node: {}, failed timesync obligation check", pubkey);
             result.timesync_status = false;
         }
+        const auto l2_threshold = m_l2_height - netconf.L2_TRACKER_SAFE_BLOCKS;
+        if (check_l2_height && l2_height < l2_threshold) {
+            log::info(
+                    logcat,
+                    "Service Node: {}, failed l2 height check. Node L2 height: {}, Threshold L2 "
+                    "height: {}",
+                    pubkey,
+                    l2_height,
+                    l2_threshold);
+            result.recent_l2_height = false;
+        }
```

Comment:
This needs a HF check, but *also* a buffer: i.e. this rule should start applying `netconf.BLOCKS_IN(netconf.UPTIME_PROOF_VALIDITY)` blocks *after* the HF20 activation block (so that we don't start failing nodes right after HF20 just because we haven't sent out a valid HF20 proof yet).